### PR TITLE
cpu/esp32: changes for toolchain update to gcc version 8.4.0

### PR DESCRIPTION
**Pull request by gschorcht**
### Contribution description

This PR provides all changes that are necessary to compile ESP32 code with the current toolchain version (gcc version 5.2.0) and the newes toolchain version (gcc 8.4.0).

This is the first step of the migration of the ESP32 port to the newest ESP-IDF SDK version 4.4. With the changes in this PR the ESP32 code can be compiled with the current toolchain version and the newest toolchain version as well. This will make it possible to switch the toolchain version without having the dependency betwenn the RIOT release and the toolchain version.

### Testing procedure

Compilation and test should succeed in CI.

### CODE REVIEW

File: cpu/esp32/Makefile.dep
Comment (benpicco) [inline cpu/esp32/Makefile.dep:11]: Is `newlib_nano` no longer supported?

File: cpu/esp_common/periph/flash.c
Comment (benpicco) [inline cpu/esp_common/periph/flash.c:465]: Why `ARRAY_SIZE` for `strncpy`?

File: cpu/esp32/Makefile.dep
Comment (gschorcht) [reply cpu/esp32/Makefile.dep:11]: gcc 8.4.0 for xtensa uses nanaospecs but doesn't have a separate include dir `newlib-nano`. Therefore the test in `makefiles/libc/newlib.mk` for this directory fails and the compilation aborts. Therefore `newlib-nano` can't be used anymore if we don't allow that this directory doesn't exist.

File: cpu/esp_common/periph/flash.c
Comment (gschorcht) [reply cpu/esp_common/periph/flash.c:465]: Otherwise, you get a compilation error:
```
cpu/esp_common/periph/flash.c:465:70: error: argument to 'sizeof' in 'strncpy' call is the same expression as the source; did you mean to use the size of the destination? [-Werror=sizeof-pointer-memaccess]
                 strncpy(part->label, (const char*)info->label, sizeof(info->label));
```

Comment (gschorcht) [reply cpu/esp_common/periph/flash.c:465]: `info->label` fact it is an character array.

Comment (benpicco) [reply cpu/esp_common/periph/flash.c:465]: Ah but the warning is correct. You are copying into `part->label`, so you should use `sizeof(part->label)`.

Comment (gschorcht) [reply cpu/esp_common/periph/flash.c:465]: Oh yeah, of course.

File: cpu/esp32/Makefile.dep
Comment (gschorcht) [reply cpu/esp32/Makefile.dep:11]:
```
makefiles/libc/newlib.mk:98: *** USE_NEWLIB_NANO==1 but nano include folder not found!
```

Comment (gschorcht) [reply cpu/esp32/Makefile.dep:11]: I already thought about to change `newlib.mk`.https://github.com/RIOT-OS/RIOT/blob/2520b5c608d393947b76b3ca8329b558c1c59dc6/makefiles/libc/newlib.mk#L97-L99
This check was introduced with PR #11145 to avoid `-isystem -I...` if the directory is not found. This assumes that a toolchain providing `newlib-nano` has such a separate include directory for the nano version of `newlib.h`. This is at least true for ARM and RISC-V.

Although `crosstool-NG` for ESP32 generates both `libc.a` and `libc_nano.a` and uses the `nano.specs`, it does not generate separate versions for `newlib.h`. The reason is probably that `crosstool-NG` for ESP32 is maintained by the developers of the `ESP-IDF` SDK which only uses the nano version of the `newlib`.

A possible solution would be to make the setting of `-isystem` dependent on whether the `newlib-nano` directory was found. This would require that a toolchain which only provides the nano version of the `newlib` is accepted.
```diff
-  ifeq (,$(NEWLIB_NANO_INCLUDE_DIR))
-    $(error USE_NEWLIB_NANO==1 but nano include folder not found!)
+  ifneq (,$(NEWLIB_NANO_INCLUDE_DIR))
+     # newlib-nano overrides newlib.h and its include dir should therefore go before
+     # the regular system include dirs.
+     INCLUDES := -isystem $(NEWLIB_NANO_INCLUDE_DIR) $(INCLUDES)
   endif
-
-  # newlib-nano overrides newlib.h and its include dir should therefore go before
-  # the regular system include dirs.
-  INCLUDES := -isystem $(NEWLIB_NANO_INCLUDE_DIR) $(INCLUDES)
 endif
```

Comment (benpicco) [reply cpu/esp32/Makefile.dep:11]: How about

```patch
--- a/makefiles/libc/newlib.mk
+++ b/makefiles/libc/newlib.mk
@@ -94,7 +94,7 @@ ifeq (1,$(USE_NEWLIB_NANO))
                                                     $(NEWLIB_INCLUDE_DIR)/newlib/nano \
                                                     $(NEWLIB_INCLUDE_DIR)/nano))
 
-  ifeq (,$(NEWLIB_NANO_INCLUDE_DIR))
+  ifeq (,$(NEWLIB_NANO_INCLUDE_DIR) $(NEWLIB_NANO_NO_INCLUDE_DIR))
     $(error USE_NEWLIB_NANO==1 but nano include folder not found!)
   endif
```

and setting `NEWLIB_NANO_NO_INCLUDE_DIR` to 1 for esp32?

Or does it not make a differnce and newlib-nano is always used either way?

Comment (gschorcht) [reply cpu/esp32/Makefile.dep:11]: This wouldn't be correct, because the line following 
https://github.com/RIOT-OS/RIOT/blob/2520b5c608d393947b76b3ca8329b558c1c59dc6/makefiles/libc/newlib.mk#L103
would then still try to add `-isystem $(NEWLIB_NANO_INCLUDE_DIR)` which isn't defined and would lead exactly to that problem the check should prevent.

For ESP32 we have only two options, either we compile the toolchain without `newlib_nano` and the standard version is always used or we compile the toolchain with `newlib_nano`, but then it has to be used always because only be one version of the `newlib.h` can exist.

I have provided my proposal as PR #17553, let's see what @kaspar30 says.

Comment (benpicco) [reply cpu/esp32/Makefile.dep:11]: Ok, then please squash

**Review by benpicco — APPROVED** [review]

File: cpu/esp32/Makefile.dep
Comment (gschorcht) [reply cpu/esp32/Makefile.dep:11]: I PR #17553 is OK, I will drop commit e7caa97. It might make sense to wait for it before squashing.

**Comment by gschorcht** [comment]
The PR has been updated as follows:
- rebased after the PR #17553 has been merged
- reverts ce10f36932e0cb72a69178dfab7723c62e651110 (a293d175ae3dad2d9961506b728dba1ac01dce0c)
- `struct __lock` size fixed (20e2841aa8171991a640d353bdb540bef8cf5eea)
- `static_assert` added to ensure that `struct __lock` can hold recursive mutex (20e2841aa8171991a640d353bdb540bef8cf5eea)

**Comment by benpicco** [comment]
Please squash!
(and drop 169e5dc9f82397db8650ced0618cf3016cf5d7f5)

**Comment by benpicco** [comment]
Just the usual, unrelated test failures

 - tests/pkg_edhoc_c/esp32-wroom-32:gnu
 - examples/twr_aloha/samr21-xpro:gnu
 - tests/pkg_mbedtls/samr21-xpro:gnu

**Comment by gschorcht** [comment]
Thanks for reviewing and merging.

**Comment by gschorcht** [comment]
Should we backport this PR and PR #17553 to the release branch. The next step of the migration to ESP-IDF 4.4 and the support of different ESP32 variants would be an upgrade from `riotdocker` to gcc 8.4.0. Only after that I can provide my already working port for ESP-IDF 4.4 as a PR. However, the upgrade to gcc 8.4.0 is only possible if these two PRs are part of the release.

I have here a fully functional port to ESP-IDF 4.4 that no longer contains vendor code and compiles the required vendor code from the ESP-IDF source code. I just need to clean up the port a little bit. So it would be a bit frustrating if we would have to wait for 2022.04 release before we can continue :wink: 

@kaspar030 What do you thinkg as the maintainer of `riotdocker`.

**Pull request merged**
